Novaflower and nettle can now properly burn/stun you

Caldony opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
fixes https://github.com/tgstation/tgstation/issues/51143#issue-619694704

fixes novaflower and deathnettle no longer having their attack() proc trigger

- what caused the bug?
the answer at its core is simple, the plant would only do their effect if the parent attack proc returned false, which it did not, ever, because it is physically unable to.

- how was the bug fixed?
yeah that's the weird part. You see the attack() proc only triggers when it's on LIVING things, however we need the nettle/novaflower to lose force whenever they attack anything, which is why afterattack is the thing in charge of making them lose force.
why not just move everything to afterattack then? And that is an option, however I settled for another, arguably much shittier solution.
see the ONLY thing in attack proc that can stop you from attacking early is the pacifism trait. and since the only reason you could probably have to not burn/stun someone is being a pacifist, I hardcoded it so it checks for pacifism, and if the guy isn't a pacifist it will stun/burn succesfully.

also you can force people to eat deathnettle on help intent so err, that's a "feature"

the more I'm looking at this the more I realised that yeeting everything into afterattack was the best and simpler fix, but eh.

## Why It's Good For The Game
put witty comment here about bugs being bad here

## Changelog
:cl:
fix: Novaflower and deathnettle will now properly burn/stun their target
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Caldony committed:
novaflower more like ho god ho fuck I'm on fire

Caldony committed:
ho god this is bad holy shit

Caldony committed:
aaaaaaaaaa

ExcessiveUseOfCobblestone commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -236,9 +236,10 @@
+	. = ..()
 	if(isliving(M))
+		if(HAS_TRAIT(user, TRAIT_PACIFISM))
+			return
```
inheritance has already ran and this value will reflect inheritance rather than being a stop in the attack chain (unless inheritance was stopped).

Is that what you want?

Caldony replied:
from what I've tested, without this it will not attack but put people on fire anyway if you are a pacifist, if you have a better fix please tell me since I'm not used to deal with traits.

ExcessiveUseOfCobblestone replied:
imagine this as `return .`

if inheritance is true, this will return true. I'm asking if that is what you want or are you looking for something like `return FALSE` here.

Caldony replied:
I still don't really get what you mean by "inheritance" is the problem

Caldony committed:
breaks the attack chain early for pacifism

guys I have no idea how to code please help

ExcessiveUseOfCobblestone commented:
inheritance is ..()

ExcessiveUseOfCobblestone commented on code/modules/hydroponics/grown/nettle.dm:
```diff
@@ -105,15 +105,17 @@
+	. = ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		return TRUE
+	if(isliving(M) && istype(M,/mob/living/carbon))
```
carbon is already deeper than living

ExcessiveUseOfCobblestone replied:
theres also a helper for carbon

ExcessiveUseOfCobblestone commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -235,15 +235,17 @@
+	. = ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		return TRUE
+	if(isliving(M) && istype(M,/mob/living/carbon))
```
same

Caldony committed:
replace isliving() with iscarbon()

I only left isliving() because it was in the original, I thought it checked if the target was alive or something

ShizCalev commented on code/modules/hydroponics/grown/nettle.dm:
```diff
@@ -105,15 +105,17 @@
+/obj/item/reagent_containers/food/snacks/grown/nettle/death/attack(mob/living/M, mob/user)
+	. = ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		return TRUE
```
return .

Caldony replied:
returning . will not interrupt the attack chain early, which is what we want. the only way to do that is to return TRUE

ShizCalev commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -235,15 +235,17 @@
+/obj/item/grown/novaflower/attack(mob/living/M, mob/user)
+	. = ..()
+	if(HAS_TRAIT(user, TRAIT_PACIFISM))
+		return TRUE
```
return .

Caldony replied:
same thing as above, returning . won't stop the attack chain early

ShizCalev commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -235,15 +235,17 @@
+		if(burned.IgniteMob())
+			message_admins("[ADMIN_LOOKUPFLW(user)] set [ADMIN_LOOKUPFLW(burned)] on fire with [src] at [AREACOORD(user)]")
+			log_game("[key_name(user)] set [key_name(burned)] on fire with [src] at [AREACOORD(user)]")
 
```
should return TRUE now.

Caldony replied:
returning TRUE now will mean it won't reach after attack, meaning that putting people on fire won't use any petals.

ShizCalev commented on code/modules/hydroponics/grown/nettle.dm:
```diff
@@ -105,15 +105,17 @@
-		M.drop_all_held_items()
+			target.Unconscious(force / 0.3)
+			target.Paralyze(force / 0.75)
+		target.drop_all_held_items()
```
after doing a thing return true

Caldony replied:
no see the problem with this is that if we return TRUE here when it succeeds, it won't reach afterattack, which means it won't lose petal when it DOES hit something.

ExcessiveUseOfCobblestone commented on code/modules/hydroponics/grown/nettle.dm:
```diff
@@ -105,15 +105,17 @@
+		var/mob/living/carbon/target = M
+		to_chat(target, "<span class='danger'>You are stunned by the powerful acid of [src]!</span>")
+		log_combat(user, target, "attacked", src)
 
-		M.adjust_blurriness(force/7)
+		target.adjust_blurriness(force/7)
 		if(prob(20))
-			M.Unconscious(force / 0.3)
-			M.Paralyze(force / 0.75)
-		M.drop_all_held_items()
+			target.Unconscious(force / 0.3)
+			target.Paralyze(force / 0.75)
+		target.drop_all_held_items()
```
early return

ExcessiveUseOfCobblestone reviewed (changes requested):
```DM
if(thing)
    do do do
// should be
if(!thing)
    return
do do do

Caldony committed:
early return moment

technically this was part of the original code which is why I never bothered trying to change the actual structure that didn't affect the bug directly, but I guess we might as well while we're at it.

ExcessiveUseOfCobblestone commented:
if you do not make changes do not resolve the review, just post why you aren't making the change and leave it open.

Caldony committed:
death nettle now says it stuns you when it actually stuns you

doesn't make sense for the death nettle to tell you you'r stunned every time even when the prop(20) doesn't trigger

ExcessiveUseOfCobblestone commented:
@ShizCalev since I forget how the request review function notifies you.

Doosteroni commented:
Staaaaaaale

Caldony commented:
whatever if shyz doesn't want to come back to review it, and apparently no maintainer can be bothered to close it/merge it themselves I guess I'll do it myself, it's like 5 line change it's not that big of a deal but I guess I'm slightly ticked off that we're going to keep a bug because of it

ExcessiveUseOfCobblestone commented:
why would you close your own PR before it autoclosed because of staling 🥴
